File: src/main/python/modular/compute/refinement.py
from collections import defaultdict, deque
from typing import Any
from modular.tree.RootedForest import RootedForest, Node
from modular.compute.MDComputeNode import MDComputeNode, OperationType, SplitDirection

__all__ = ['refine']

VertexId = Any


def trace(message: str, *args: object) -> None:
    print(f'[TRACE] {message}', *args)


def refine(
    tree: RootedForest[MDComputeNode],
    vertex_nodes: dict[VertexId, Node[MDComputeNode]],
    alpha_list: dict[VertexId, set[VertexId]],
    prob: Node[MDComputeNode]
) -> None:
    number_by_comp(prob)
    number_by_tree(prob)

    # trace('comp:', {v: v.data.comp_number for v in prob.bfs_nodes()})
    # trace('tree:', {v: v.data.tree_number for v in prob.bfs_nodes()})

    for v in prob.get_leaves():
        refine_with(tree, vertex_nodes, alpha_list, v.data.vertex, prob.data.vertex)
        # trace(f'refined at {v}: tree={repr(tree)}')

# ===============================================================================
#    Set up
# ===============================================================================


def number_by_comp(prob: Node[MDComputeNode]) -> None:
    """Updtaes the component number for each node in the given problem subtree."""
    comp_number = 0
    pivot = prob.data.vertex
    op_type: OperationType = OperationType.SERIES

    for c in prob.get_children():
        if c.data.vertex == pivot:
            op_type = OperationType.PARALLEL

        if c.data.op_type == op_type:
            for x in c.get_children():
                for y in x.dfs_reverse_preorder_nodes():
                    y.data.comp_number = comp_number
                comp_number += 1
        else:
            for y in c.dfs_reverse_preorder_nodes():
                y.data.comp_number = comp_number
            comp_number += 1


def number_by_tree(prob: Node[MDComputeNode]) -> None:
    """Updtaes the tree number for each node in the given problem subtree."""
    tree_number = 0
    for c in prob.get_children():
        for y in c.dfs_reverse_preorder_nodes():
            y.data.tree_number = tree_number
        tree_number += 1


# ===============================================================================
#    Utilities
# ===============================================================================
def is_root_operator(node: Node[MDComputeNode]) -> bool:
    return node.parent is None or not node.parent.data.is_operation_node()


# ===============================================================================
#    Marking split types
# ===============================================================================
def mark_ancestors_by_split(node: Node[MDComputeNode], split_type: SplitDirection) -> None:
    """Adds the given mark to all of the node's ancestors."""
    p = node.parent
    while p is not None:
        if p.data.is_problem_node():
            break  # passed the operation root
        if p.data.is_split_marked(split_type):
            add_split_mark(p, split_type, should_recurse=True)
            break
        add_split_mark(p, split_type, should_recurse=True)
        p = p.parent


def add_split_mark(node: Node[MDComputeNode], split_type: SplitDirection, should_recurse: bool) -> None:
    """Add the given mark to the node and possibly its children."""
    if not node.data.is_split_marked(split_type):
        if node.parent is not None and node.parent.data.is_operation_node():
            node.parent.data.increment_num_split_children(split_type)
        node.data.set_split_mark(split_type)

    if not should_recurse or node.data.op_type != OperationType.PRIME:
        return

    if node.number_of_children() == node.data.get_num_split_children(split_type):
        return  # split type is already set to all children

    for c in node.get_children():
        if not c.data.is_split_marked(split_type):
            node.data.increment_num_split_children(split_type)
            c.data.set_split_mark(split_type)


# ===============================================================================
#    Get max subtrees
# ===============================================================================
def is_parent_fully_charged(x: Node[MDComputeNode]) -> bool:
    if is_root_operator(x):
        return False
    p = x.parent
    assert p is not None
    return p.number_of_children() == p.data.num_marks


def get_max_subtrees(leaves: list[Node[MDComputeNode]]) -> list[Node[MDComputeNode]]:
    """
    Finds the set of maximal subtrees such that the leaves of each subtree are
    subsets of the given leaf set.
    """
    full_charged = list(leaves)
    charged: list[Node[MDComputeNode]] = []

    # charging
    idx = 0
    while idx < len(full_charged):
        x = full_charged[idx]
        if not is_root_operator(x):
            p = x.get_parent()
            if p.data.num_marks == 0:
                charged += [p]
            p.data.add_mark()

            if p.number_of_children() == p.data.num_marks:
                # fully charged
                full_charged += [p]
        idx += 1

    # discharging
    ret: list[Node[MDComputeNode]] = [x for x in full_charged if not is_parent_fully_charged(x)]

    # cleaning
    for x in charged:
        x.data.clear_marks()
    return ret


# ===============================================================================
#    Group sibling nodes
# ===============================================================================
def group_sibling_nodes(
    tree: RootedForest[MDComputeNode],
    nodes: list[Node[MDComputeNode]]
) -> list[tuple[Node[MDComputeNode], bool]]:
    """
    Precondition: `nodes` must be maximal subtrees (no node is an ancestor of another).
    """
    parents: list[Node[MDComputeNode]] = []
    sibling_groups: list[tuple[Node[MDComputeNode], bool]] = []

    for node in nodes:
        if is_root_operator(node):
            # (1) roots of trees
            sibling_groups += [(node, False)]
        else:
            tree.make_first_child(node)
            p = node.parent
            assert p is not None

            if p.data.num_marks == 0:
                parents += [p]
            p.data.add_mark()

    for p in parents:
        # there must be at least one mark
        c = p.first_child
        assert c is not None

        if p.data.num_marks == 1:
            # (2) non-root nodes without siblings
            sibling_groups += [(c, False)]
        else:
            # (3) group sibling nodes as the children of a newly inserted node
            grouped_children = tree.create_node(p.data.copy())

            for split_type in [SplitDirection.LEFT, SplitDirection.RIGHT]:
                if grouped_children.data.is_split_marked(split_type):
                    p.data.increment_num_split_children(split_type)

            # FIX FROM ORIGINAL CODE (RecSubProblem.java line 971); bug when a child of `p` is also in `parents`
            for c in p.get_children()[:p.data.num_marks]:
                for split_type in [SplitDirection.LEFT, SplitDirection.RIGHT]:
                    if c.data.is_split_marked(split_type):
                        p.data.decrement_num_split_children(split_type)
                        grouped_children.data.increment_num_split_children(split_type)

                tree.move_to(c, grouped_children)
            tree.move_to(grouped_children, p)

            sibling_groups += [(grouped_children, grouped_children.data.op_type == OperationType.PRIME)]

        # clean marks
        p.data.clear_marks()

    return sibling_groups


# ===============================================================================
#    Subroutine
# ===============================================================================
def get_split_type(
    vertex_nodes: dict[VertexId, Node[MDComputeNode]],
    refiner: VertexId,
    pivot: VertexId,
    node: Node[MDComputeNode]
) -> SplitDirection:
    pivot_tn = vertex_nodes[pivot].data.tree_number
    refiner_tn = vertex_nodes[refiner].data.tree_number
    current_tn = node.data.tree_number
    return SplitDirection.LEFT if current_tn < pivot_tn or refiner_tn < current_tn else SplitDirection.RIGHT


def refine_one_node(
    tree: RootedForest[MDComputeNode],
    node: Node[MDComputeNode],
    split_type: SplitDirection,
    new_prime: bool,
) -> None:
    if is_root_operator(node):
        return

    p = node.parent
    new_sibling = None

    assert p is not None

    if is_root_operator(p):
        # parent is a root; must split the tree
        if split_type == SplitDirection.LEFT:
            tree.move_to_before(node, p)
        else:
            tree.move_to_after(node, p)
        for st in [SplitDirection.LEFT, SplitDirection.RIGHT]:
            if node.data.is_split_marked(st):
                p.data.decrement_num_split_children(st)

        new_sibling = p

        if p.has_only_one_child():
            tree.replace_by_children(p)
            tree.remove(p)
            new_sibling = None
        else:
            assert not p.is_leaf()

    elif p.data.op_type != OperationType.PRIME:
        # parent is not a root or PRIME
        assert p.parent is not None
        assert p.number_of_children() > 1, 'p should have multiple children at this point'

        replacement = tree.create_node(p.data.copy())
        tree.replace(p, replacement)
        tree.move_to(node, replacement)
        tree.move_to(p, replacement)
        new_sibling = p

        for st in [SplitDirection.LEFT, SplitDirection.RIGHT]:
            if node.data.is_split_marked(st):
                p.data.decrement_num_split_children(st)
                replacement.data.increment_num_split_children(st)
            if p.data.is_split_marked(st):
                replacement.data.increment_num_split_children(st)

    add_split_mark(node, split_type, should_recurse=new_prime)
    mark_ancestors_by_split(node, split_type)

    if new_sibling:
        # non-prime or a new root; safe to set should_recurse=True
        add_split_mark(new_sibling, split_type, should_recurse=True)


def refine_with(
    tree: RootedForest[MDComputeNode],
    vertex_nodes: dict[VertexId, Node[MDComputeNode]],
    alpha_list: dict[VertexId, set[VertexId]],
    refiner: VertexId,
    pivot: VertexId
) -> None:
    subtree_roots = get_max_subtrees([vertex_nodes[x] for x in alpha_list[refiner]])
    sibling_groups = group_sibling_nodes(tree, subtree_roots)
    # trace(f'refiner={refiner}, subtree_roots={subtree_roots}, sibling_groups={sibling_groups}, tree={tree}')

    for current, new_prime in sibling_groups:
        # determine the split type
        split_type = get_split_type(vertex_nodes, refiner, pivot, current)
        # trace(f'before: refiner={refiner}, node={current}, tree={tree}')
        refine_one_node(tree, current, split_type, new_prime)
        # trace(f'after : refiner={refiner}, node={current}, tree={tree}')

        # sanity check
        # for r in tree.roots:
        #     for node in r.dfs_reverse_preorder_nodes():
        #         if not node.data.is_operation_node():
        #             assert node.data.num_left_split_children == 0
        #             assert node.data.num_right_split_children == 0
        #         else:
        #             lact = node.data.num_left_split_children
        #             ract = node.data.num_right_split_children
        #             lexp = sum(1 if c.data.is_split_marked(SplitDirection.LEFT) else 0 for c in node.get_children())
        #             rexp = sum(1 if c.data.is_split_marked(SplitDirection.RIGHT) else 0 for c in node.get_children())
        #             assert lact == lexp, f'node={node}, expect={lexp}, actual={lact}'
        #             assert ract == rexp, f'node={node}, expect={rexp}, actual={lact}'

File: src/main/cpp/modular/compute/refinement.cpp
#include "MDSolver.hpp"

namespace modular {
namespace compute {
namespace impl {

// split directions: left or right
std::vector<SplitDirection> const DIRS = {SplitDirection::LEFT, SplitDirection::RIGHT};

//================================================================================
//    Set up
//================================================================================
/**
 * @brief Updates the component number for each node in tree given problem subtree.
 *
 * @param tree tree
 * @param prob problem root
 */
static void number_by_comp(CompTree &tree, int prob) {
  int comp_number = 0;
  int pivot = tree[prob].data.vertex;
  Operation op_type = Operation::SERIES;

  for (auto c : tree.get_children(prob)) {
    if (c == pivot) op_type = Operation::PARALLEL;

    if (tree[c].data.op_type == op_type) {
      for (auto x : tree.get_children(c)) {
        for (auto y : tree.dfs_reverse_preorder_nodes(x)) tree[y].data.comp_number = comp_number;
        ++comp_number;
      }
    } else {
      for (auto y : tree.dfs_reverse_preorder_nodes(c)) tree[y].data.comp_number = comp_number;
      ++comp_number;
    }
  }
}

static void number_by_tree(CompTree &tree, int prob) {
  int tree_number = 0;
  for (auto c : tree.get_children(prob)) {
    for (auto y : tree.dfs_reverse_preorder_nodes(c)) tree[y].data.tree_number = tree_number;
    ++tree_number;
  }
}

//================================================================================
//    Utilities
//================================================================================
static bool is_root_operator(CompTree const &tree, int index) {
  return tree[index].is_root() || !tree[tree[index].parent].data.is_operation_node();
}

//================================================================================
//    Marking split types
//================================================================================
/**
 * @brief Adds the given mark to the node and possibly its children.
 *
 * @param tree tree
 * @param index index
 * @param split_type split type
 * @param should_recurse true if it needs to recurse to children
 */
static void add_split_mark(CompTree &tree, int index, SplitDirection split_type, bool should_recurse, util::Profiler *prof) {
  if (!tree[index].data.is_split_marked(split_type)) {
    auto p = tree[index].parent;  // must be a valid node
    // increment the counter if the parent is an operation node
    if (tree[p].data.is_operation_node()) tree[p].data.increment_num_split_children(split_type);
    tree[index].data.set_split_mark(split_type);
  }

  if (!should_recurse || tree[index].data.op_type != Operation::PRIME) return;

  // split type is already set to all children
  if (tree[index].number_of_children() == tree[index].data.get_num_split_children(split_type)) {
    // PROF(util::pcount(prof, "add_split_mark(): early return"));
    return;
  }

  // PROF(util::pcount(prof, "add_split_mark(): proc child"));
  for (auto c = tree[index].first_child; tree.is_valid(c); c = tree[c].right) {
    if (!tree[c].data.is_split_marked(split_type)) {
      tree[index].data.increment_num_split_children(split_type);
      tree[c].data.set_split_mark(split_type);
    }
  }
}

/**
 * @brief Adds the given mark to all of this node's ancestors.
 * @param split_type mark to be added
 */
static void mark_ancestors_by_split(CompTree &tree, int index, SplitDirection split_type, util::Profiler *prof) {
  for (auto p = tree[index].parent;; p = tree[p].parent) {
    if (tree[p].data.is_problem_node()) break;
    if (tree[p].data.is_split_marked(split_type)) {
      // split type is already set to p but we need to take care of its children
      add_split_mark(tree, p, split_type, true, prof);
      break;
    }
    add_split_mark(tree, p, split_type, true, prof);
  }
}

//================================================================================
//    Get max subtrees
//================================================================================
static bool is_parent_fully_charged(CompTree const &tree, int x) {
  if (is_root_operator(tree, x)) return false;
  auto p = tree[x].parent;
  return tree[p].number_of_children() == tree[p].data.number_of_marks();
}

/**
 * @brief Finds the set of maximal subtrees such that the leaves of
 * each subtree are subsets of the given leaf set.
 *
 * @param leaves
 * @return std::list<NodeP>
 */
static std::vector<int> get_max_subtrees(CompTree &tree, VI const &leaves) {
  std::vector<int> full_charged(leaves);
  std::vector<int> charged;

  // charging
  for (std::size_t i = 0; i < full_charged.size(); ++i) {
    auto x = full_charged[i];
    if (is_root_operator(tree, x)) continue;

    auto p = tree[x].parent;
    if (!tree[p].data.is_marked()) charged.push_back(p);
    tree[p].data.add_mark();

    if (tree[p].data.num_marks == tree[p].number_of_children()) {
      // fully charged
      full_charged.push_back(p);
    }
  }

  // discharging
  std::vector<int> ret;
  for (auto x : full_charged) {
    if (!is_parent_fully_charged(tree, x)) ret.push_back(x);
  }
  for (auto x : charged) tree[x].data.clear_marks();
  return ret;
}

//================================================================================
//    Group sibling nodes
//================================================================================
std::vector<std::pair<int, bool>> group_sibling_nodes(CompTree &tree, std::vector<int> const &nodes) {
  std::vector<int> parents;
  std::vector<std::pair<int, bool>> sibling_groups;

  for (auto node : nodes) {
    if (is_root_operator(tree, node)) {
      // (1) the roots of trees
      sibling_groups.push_back({node, false});
    } else {
      tree.make_first_child(node);
      auto p = tree[node].parent;

      if (!tree[p].data.is_marked()) parents.push_back(p);
      tree[p].data.add_mark();
    }
  }

  for (auto p : parents) {
    // there must be at least one mark
    auto c = tree[p].first_child;
    auto num_marks = tree[p].data.number_of_marks();

    if (num_marks == 1) {
      // (2) the non-root nodes without siblings
      sibling_groups.push_back({c, false});
    } else {
      // (3) group sibling nodes as children of a new node inserted in their place.
      auto grouped_children = tree.create_node(tree[p].data);
      tree[grouped_children].data.clear_marks();

      for (auto st : DIRS) {
        if (tree[grouped_children].data.is_split_marked(st)) tree[p].data.increment_num_split_children(st);
      }

      auto c = tree[p].first_child;
      for (int i = 0; i < num_marks; ++i) {
        auto nxt = tree[c].right;
        tree.move_to(c, grouped_children);

        for (auto st : DIRS) {
          if (tree[c].data.is_split_marked(st)) {
            tree[p].data.decrement_num_split_children(st);
            tree[grouped_children].data.increment_num_split_children(st);
          }
        }
        c = nxt;
      }
      tree.move_to(grouped_children, p);

      sibling_groups.push_back({grouped_children, tree[grouped_children].data.op_type == Operation::PRIME});
    }
    tree[p].data.clear_marks();
  }

  // TRACE("return: %s\n", cstr(sibling_groups));
  return sibling_groups;
}

//================================================================================
//    Subroutine
//================================================================================
static SplitDirection get_split_type(CompTree &tree, int index, VertexID refiner, VertexID pivot) {
  auto pivot_tn = tree[pivot].data.tree_number;
  auto refiner_tn = tree[refiner].data.tree_number;
  auto current = tree[index].data.tree_number;
  return current < pivot_tn || refiner_tn < current ? SplitDirection::LEFT : SplitDirection::RIGHT;
}

static void refine_one_node(CompTree &tree, int index, SplitDirection split_type, bool new_prime, util::Profiler *prof) {
  TRACE("refining tree=%s, index=%d, split_type=%d, new_prime=%d", tree.to_string(index).c_str(), index, split_type, new_prime);
  if (is_root_operator(tree, index)) return;

  auto p = tree[index].parent;
  int new_sibling = -1;

  if (is_root_operator(tree, p)) {
    // PROF(util::pstart(prof, "refine_one_node: root", 0));
    // parent is a root; must split there
    if (split_type == SplitDirection::LEFT) {
      tree.move_to_before(index, p);
    } else {
      tree.move_to_after(index, p);
    }

    for (auto st : DIRS) {
      if (tree[index].data.is_split_marked(st)) tree[p].data.decrement_num_split_children(st);
    }

    new_sibling = p;

    if (tree[p].has_only_one_child()) {
      tree.replace_by_children(p);
      tree.remove(p);
      new_sibling = -1;
    }
    // PROF(util::pstop(prof, "refine_one_node: root", 0));
  } else if (tree[p].data.op_type != Operation::PRIME) {
    // PROF(util::pstart(prof, "refine_one_node: non-root", 0));
    // parent is not a root or PRIME
    auto replacement = tree.create_node(tree[p].data);
    tree.replace(p, replacement);
    tree.move_to(index, replacement);
    tree.move_to(p, replacement);
    new_sibling = p;

    for (auto st : DIRS) {
      if (tree[index].data.is_split_marked(st)) {
        tree[p].data.decrement_num_split_children(st);
        tree[replacement].data.increment_num_split_children(st);
      }
      if (tree[p].data.is_split_marked(st)) tree[replacement].data.increment_num_split_children(st);
    }
    // PROF(util::pstop(prof, "refine_one_node: non-root", 0));
  }

  // PROF(util::pstart(prof, "add_split_mark()", 1));
  add_split_mark(tree, index, split_type, new_prime, prof);
  // PROF(util::pstop(prof, "add_split_mark()", 1));

  // PROF(util::pstart(prof, "mark_ancestors_by_split()"));
  mark_ancestors_by_split(tree, index, split_type, prof);
  // PROF(util::pstop(prof, "mark_ancestors_by_split()"));

  if (new_sibling >= 0) {
    // non-prime or a new root; safe to set should_recurse=true
    // PROF(util::pstart(prof, "add_split_mark()", 2));
    add_split_mark(tree, new_sibling, split_type, true, prof);
    // PROF(util::pstop(prof, "add_split_mark()", 2));
  }
}

static void refine_with(CompTree &tree, VVV const &alpha_list, VertexID refiner, VertexID pivot, util::Profiler *prof) {
  // PROF(util::pstart(prof, "get_max_subtrees()"))
  auto subtree_roots = get_max_subtrees(tree, alpha_list[refiner]);
  // PROF(util::pstop(prof, "get_max_subtrees()"))

  // PROF(util::pstart(prof, "group_sibling_nodes()"))
  auto sibling_groups = group_sibling_nodes(tree, subtree_roots);
  // PROF(util::pstop(prof, "group_sibling_nodes()"))

  TRACE("alpha[%d]: %s", refiner, util::to_string(alpha_list[refiner]).c_str());
  TRACE("subtree_roots: %s", util::to_string(subtree_roots).c_str());
  TRACE("sibling_groups: %s, tree=%s", util::to_string(sibling_groups).c_str(), tree.to_string(tree[pivot].parent).c_str());

  // PROF(util::pstart(prof, "refine_with: main loop"))
  for (auto &x : sibling_groups) {
    // PROF(util::pstart(prof, "get_split_type()"))
    auto split_type = get_split_type(tree, x.first, refiner, pivot);
    // PROF(util::pstop(prof, "get_split_type()"))

    // PROF(util::pstart(prof, "refine_one_node()"))
    refine_one_node(tree, x.first, split_type, x.second, prof);
    // PROF(util::pstop(prof, "refine_one_node()"))
  }
  // PROF(util::pstop(prof, "refine_with: main loop"))
}

//================================================================================
//    Refinement
//================================================================================
void refine(CompTree &tree, VVV const &alpha_list, int prob, std::vector<int> &leaves, util::Profiler *prof) {
  TRACE("start: %s", tree.to_string(prob).c_str());
  // PROF(util::pstart(prof, "refinement:refine()"))

  // PROF(util::pstart(prof, "refinement:number_by_comp()"))
  number_by_comp(tree, prob);
  // PROF(util::pstop(prof, "refinement:number_by_comp()"))

  // PROF(util::pstart(prof, "refinement:number_by_tree()"))
  number_by_tree(tree, prob);
  // PROF(util::pstop(prof, "refinement:number_by_tree()"))

  // PROF(util::pstart(prof, "refinement:refine_with()"));
  for (auto v : leaves) {
    refine_with(tree, alpha_list, v, tree[prob].data.vertex, prof);
    TRACE("refined at %d: tree=%s", v, tree.to_string(prob).c_str())
  }
  // PROF(util::pstop(prof, "refinement:refine_with()"));

  // PROF(util::pstop(prof, "refinement:refine()"))
  TRACE("finish: %s", tree.to_string(prob).c_str());
}

}  // namespace impl
}  // namespace compute
}  // namespace modular
